File: src/nostr/nip_47_test.rs
use super::*;

    #[test]
    fn test_nwc_uri_from_uri() {
        let uri = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101";
        let nwc_uri = NwcUri::from_uri(uri).unwrap();
        assert_eq!(
            nwc_uri.wallet_pubkey,
            "1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f"
        );
        assert_eq!(
            nwc_uri.secret,
            "0101010101010101010101010101010101010101010101010101010101010101"
        );
    }

    #[test]
    fn test_nwc_client_roundtrip() {
        let uri_str = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101";
        let nwc_uri = NwcUri::from_uri(uri_str).unwrap();
        let client = NwcClient::new(nwc_uri).unwrap();

        let payload = serde_json::json!({"test": "data"});
        let encrypted = client.encrypt(&payload).unwrap();
        let decrypted = client.decrypt(&encrypted).unwrap();
        let decrypted_json: Value = serde_json::from_str(&decrypted).unwrap();
        assert_eq!(payload, decrypted_json);

        let (event, request_id) = client
            .create_request_event(NwcMethod::MakeInvoice, payload.clone(), vec![])
            .unwrap();
        assert_eq!(event.pubkey, client.my_pubkey);
        assert_eq!(event.kind, KIND_NWC_REQUEST);

        // Test parsing back (mocking a response)
        let resp_payload = serde_json::json!({"result": {"invoice": "lnbc1..."}});
        let resp_encrypted = client.encrypt(&resp_payload).unwrap();
        let resp_event = Event {
            id: "resp_id".into(),
            pubkey: client.wallet_pubkey.clone(),
            created_at: crate::util::now(),
            kind: 23195,
            tags: vec![
                Tag::E(request_id.clone(), vec![]),
                Tag::p(&client.my_pubkey),
            ],
            content: resp_encrypted,
            sig: "sig".into(), // verify(now()) will fail in test unless we sign it properly, but we'll bypass verification for this unit test if needed or just sign it.
        };
        
        // Actually we need to sign it to pass verify(now())
        let wallet_sk_bytes = hex::decode("0101010101010101010101010101010101010101010101010101010101010101").unwrap();
        let wallet_sk_arr: [u8; 32] = wallet_sk_bytes.try_into().unwrap();
        let wallet_sk = SigningKey::from_bytes(&wallet_sk_arr).unwrap();
        
        let (resp_id, resp_id_bytes) = Event::compute_id(&client.wallet_pubkey, resp_event.created_at, resp_event.kind, &resp_event.tags, &resp_event.content).unwrap();
        let resp_sig = hex::encode(wallet_sk.sign_prehash(&resp_id_bytes).unwrap().to_bytes());
        
        let signed_resp_event = Event {
            id: resp_id,
            sig: resp_sig,
            ..resp_event
        };

        let parsed_resp = client.parse_response_event(&signed_resp_event, &request_id).unwrap();
        assert_eq!(parsed_resp, resp_payload);
    }

    #[test]
    fn test_nwc_client_nip44_roundtrip() {
        let uri_str = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101";
        let nwc_uri = NwcUri::from_uri(uri_str).unwrap();
        let mut client = NwcClient::new(nwc_uri).unwrap();
        client.encryption_method = EncryptionMethod::Nip44;

        let payload = serde_json::json!({"test": "nip44 data"});
        let encrypted = client.encrypt(&payload).unwrap();
        let decrypted = client.decrypt(&encrypted).unwrap();
        let decrypted_json: Value = serde_json::from_str(&decrypted).unwrap();
        assert_eq!(payload, decrypted_json);
    }

    #[test]
    #[should_panic(expected = "InvalidScheme")]
    fn test_uri_invalid_scheme() {
        let uri = "http://invalid.example.com?secret=0101010101010101010101010101010101010101010101010101010101010101";
        let _ = NwcUri::from_uri(uri).unwrap();
    }

    #[test]
    fn test_uri_missing_pubkey_returns_error() {
        let uri = "nostr+walletconnect://?secret=0101010101010101010101010101010101010101010101010101010101010101";
        let result = NwcUri::from_uri(uri);
        assert!(result.is_err());
    }

    #[test]
    fn test_uri_missing_secret_returns_error() {
        let uri = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f";
        let result = NwcUri::from_uri(uri);
        assert!(result.is_err());
    }

    #[test]
    fn test_client_encrypt_deterministic() {
        let uri_str = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101";
        let nwc_uri = NwcUri::from_uri(uri_str).unwrap();
        let client = NwcClient::new(nwc_uri).unwrap();

        let payload = serde_json::json!({"same": "data"});
        let encrypted1 = client.encrypt(&payload).unwrap();
        let encrypted2 = client.encrypt(&payload).unwrap();
        assert_ne!(encrypted1, encrypted2);
    }

    #[test]
    fn test_client_created_has_required_fields() {
        let uri_str = "nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101";
        let nwc_uri = NwcUri::from_uri(uri_str).unwrap();
        let client = NwcClient::new(nwc_uri).unwrap();

        assert!(!client.my_pubkey.is_empty());
    }

    #[test]
    fn test_kind_constants() {
        assert_eq!(KIND_NWC_REQUEST, 23194);
    }

    #[test]
    fn test_encryption_method_from_protocol_str() {
        assert_eq!(EncryptionMethod::from_protocol_str("nip04"), Some(EncryptionMethod::Nip04));
        assert_eq!(EncryptionMethod::from_protocol_str("nip44_v2"), Some(EncryptionMethod::Nip44));
        assert_eq!(EncryptionMethod::from_protocol_str("unknown"), None);
    }

    #[test]
    fn test_encryption_method_to_protocol_string() {
        assert_eq!(EncryptionMethod::Nip04.to_protocol_string(), "nip04");
        assert_eq!(EncryptionMethod::Nip44.to_protocol_string(), "nip44_v2");
    }

    #[test]
    fn test_nwc_uri_error_display() {
        assert!(NwcUriError::InvalidUrl("bad".into()).to_string().contains("url failure"));
        assert!(NwcUriError::InvalidScheme.to_string().contains("Invalid scheme"));
        assert!(NwcUriError::MissingPubkey.to_string().contains("Missing wallet pubkey"));
        assert!(NwcUriError::MissingSecret.to_string().contains("Missing secret"));
    }

    #[test]
    fn test_parse_wallet_info_no_encryption_tag_defaults_nip04() {
        let event = Event {
            id: "id1".into(),
            pubkey: "pk1".into(),
            created_at: 1000,
            kind: 13194,
            tags: vec![],
            content: "".into(),
            sig: "sig".into(),
        };
        let info = parse_wallet_info(&event);
        assert_eq!(info.encryption_algorithms, vec![EncryptionMethod::Nip04]);
    }

File: src/nostr/nip47.test.ts
import { describe, it, expect } from 'vitest';
import { hexEncode, hexDecode } from '../common/util.ts';
import { schnorr } from '@noble/curves/secp256k1.js';
import { Tag } from './tag.ts';
import { computeEventId } from './event.ts';
import type { Event } from './event.ts';
import {
  KIND_NWC_REQUEST,
  EncryptionMethod,
  encryptionToProtocol,
  encryptionFromProtocol,
  NwcUriError,
  parseNwcUri,
  NwcClient,
  parseWalletInfo,
} from './nip47.ts';

const TEST_WALLET_SK = '0101010101010101010101010101010101010101010101010101010101010101';
const TEST_WALLET_PK = '1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f';
const TEST_NWC_URI = 'nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f?relay=ws%3A%2F%2Flocalhost%3A8787%2F&secret=0101010101010101010101010101010101010101010101010101010101010101';

const NOW = 1700000000;
const mockClock = () => NOW;

describe('NIP-47', () => {

  describe('NwcUri', () => {
    it('parses a valid NWC URI', () => {
      const nwcUri = parseNwcUri(TEST_NWC_URI);
      expect(nwcUri.walletPubkey).toBe(TEST_WALLET_PK);
      expect(nwcUri.secret).toBe(TEST_WALLET_SK);
    });

    it('throws on invalid scheme', () => {
      const uri = 'http://invalid.example.com?secret=0101010101010101010101010101010101010101010101010101010101010101';
      expect(() => parseNwcUri(uri)).toThrow(NwcUriError);
    });

    it('throws on missing pubkey', () => {
      const uri = 'nostr+walletconnect://?secret=0101010101010101010101010101010101010101010101010101010101010101';
      expect(() => parseNwcUri(uri)).toThrow(NwcUriError);
    });

    it('throws on missing secret', () => {
      const uri = 'nostr+walletconnect://1b84c5567b126440995d3ed5aaba0565d71e1834604819ff9c17f5e9d5dd078f';
      expect(() => parseNwcUri(uri)).toThrow(NwcUriError);
    });
  });

  describe('NwcClient', () => {
    function createClient() {
      const nwcUri = parseNwcUri(TEST_NWC_URI);
      return new NwcClient(nwcUri, mockClock);
    }

    it('has required fields on construction', () => {
      const client = createClient();
      expect(client.myPubkey).toBeTruthy();
      expect(client.myPubkey.length).toBeGreaterThan(0);
    });

    it('encrypt/decrypt roundtrip (NIP-04)', async () => {
      const client = createClient();
      const payload = { test: 'data' };
      const encrypted = await client.encrypt(payload);
      const decrypted = await client.decrypt(encrypted);
      expect(JSON.parse(decrypted)).toEqual(payload);
    });

    it('encrypt/decrypt roundtrip (NIP-44)', async () => {
      const client = createClient();
      client.encryptionMethod = EncryptionMethod.Nip44;
      const payload = { test: 'nip44 data' };
      const encrypted = await client.encrypt(payload);
      const decrypted = await client.decrypt(encrypted);
      expect(JSON.parse(decrypted)).toEqual(payload);
    });

    it('produces different ciphertexts for same plaintext', async () => {
      const client = createClient();
      const payload = { same: 'data' };
      const encrypted1 = await client.encrypt(payload);
      const encrypted2 = await client.encrypt(payload);
      expect(encrypted1).not.toBe(encrypted2);
    });

    it('create request event has required fields', async () => {
      const client = createClient();
      const { event, requestId } = await client.createRequestEvent('make_invoice', {}, []);
      expect(event.pubkey).toBe(client.myPubkey);
      expect(event.kind).toBe(KIND_NWC_REQUEST);
      expect(event.id).toBe(requestId);
    });

    it('full request/response roundtrip', async () => {
      const client = createClient();
      const payload = { test: 'data' };

      const { requestId } = await client.createRequestEvent('make_invoice', payload, []);

      const respPayload = { result: { invoice: 'lnbc1...' } };
      const respEncrypted = await client.encrypt(respPayload);

      const respCreatedAt = NOW;
      const respTags = [
        Tag.e(requestId),
        Tag.p(client.myPubkey),
      ];
      const { id: respId, idBytes: respIdBytes } = computeEventId(client.walletPubkey, respCreatedAt, 23195, respTags, respEncrypted);
      const walletSkBytes = hexDecode(TEST_WALLET_SK);
      const respSigBytes = schnorr.sign(respIdBytes, walletSkBytes);
      const respSig = hexEncode(respSigBytes);

      const signedRespEvent: Event = {
        id: respId,
        pubkey: client.walletPubkey,
        createdAt: respCreatedAt,
        kind: 23195,
        tags: respTags,
        content: respEncrypted,
        sig: respSig,
      };

      const parsedResp = await client.parseResponseEvent(signedRespEvent, requestId);
      expect(parsedResp).toEqual(respPayload);
    });
  });

  describe('EncryptionMethod', () => {
    it('converts to protocol string', () => {
      expect(encryptionToProtocol(EncryptionMethod.Nip04)).toBe('nip04');
      expect(encryptionToProtocol(EncryptionMethod.Nip44)).toBe('nip44_v2');
    });

    it('converts from protocol string', () => {
      expect(encryptionFromProtocol('nip04')).toBe(EncryptionMethod.Nip04);
      expect(encryptionFromProtocol('nip44_v2')).toBe(EncryptionMethod.Nip44);
      expect(encryptionFromProtocol('unknown')).toBeNull();
    });
  });

  describe('NwcUriError', () => {
    it('has expected error messages', () => {
      expect(NwcUriError.invalidUrl('bad').message).toContain('url failure');
      expect(NwcUriError.invalidScheme().message).toContain('Invalid scheme');
      expect(NwcUriError.missingPubkey().message).toContain('Missing wallet pubkey');
      expect(NwcUriError.missingSecret().message).toContain('Missing secret');
    });
  });

  describe('parseWalletInfo', () => {
    it('defaults to NIP-04 when no encryption tag', () => {
      const event: Event = {
        id: 'id1',
        pubkey: 'pk1',
        createdAt: 1000,
        kind: 13194,
        tags: [],
        content: '',
        sig: 'sig',
      };
      const info = parseWalletInfo(event);
      expect(info.encryptionAlgorithms).toEqual([EncryptionMethod.Nip04]);
    });
  });

  describe('kind constants', () => {
    it('KIND_NWC_REQUEST is 23194', () => {
      expect(KIND_NWC_REQUEST).toBe(23194);
    });
  });
});
